fe/fe-core/src/main/java/com/starrocks/sql/analyzer/PartitionExprAnalyzer.java

```diff
@@ -14,26 +14,47 @@
 
 package com.starrocks.sql.analyzer;
 
+import com.starrocks.analysis.CastExpr;
 import com.starrocks.analysis.Expr;
 import com.starrocks.analysis.FunctionCallExpr;
+import com.starrocks.analysis.SlotRef;
 import com.starrocks.catalog.Function;
 import com.starrocks.catalog.FunctionSet;
 import com.starrocks.catalog.Type;
+import com.starrocks.common.AnalysisException;
+
+import java.util.ArrayList;
 
 public class PartitionExprAnalyzer {
 
-    public static void analyzePartitionExpr(Expr expr, Type targetColType) {
+    public static void analyzePartitionExpr(Expr expr, SlotRef partitionSlotRef) {
         if (expr instanceof FunctionCallExpr) {
             FunctionCallExpr functionCallExpr = (FunctionCallExpr) expr;
             Function builtinFunction = null;
-            if (functionCallExpr.getFnName().getFunction().equalsIgnoreCase(FunctionSet.DATE_TRUNC)) {
+            Type targetColType = partitionSlotRef.getType();
+            String functionName = functionCallExpr.getFnName().getFunction();
+            if (functionName.equalsIgnoreCase(FunctionSet.DATE_TRUNC)) {
                 Type[] dateTruncType = {Type.VARCHAR, targetColType};
                 builtinFunction = Expr.getBuiltinFunction(functionCallExpr.getFnName().getFunction(),
                         dateTruncType, Function.CompareMode.IS_IDENTICAL);
-            } else if (functionCallExpr.getFnName().getFunction().equalsIgnoreCase(FunctionSet.TIME_SLICE)) {
+            } else if (functionName.equalsIgnoreCase(FunctionSet.TIME_SLICE)) {
                 Type[] timeSliceType = {targetColType, Type.INT, Type.VARCHAR, Type.VARCHAR};
                 builtinFunction = Expr.getBuiltinFunction(functionCallExpr.getFnName().getFunction(),
                         timeSliceType, Function.CompareMode.IS_IDENTICAL);
+            } else if (functionName.equalsIgnoreCase(FunctionSet.SUBSTR) ||
```

Comment:
is it necessary to make it more universal

Reply:
in the future PR.